File: installer/opt_components.py
# -*- coding: utf-8 -*-
"""可选组件：检测 + 在线下载官方安装器 + 静默安装（不内置静态资源）
支持：Node.js / Python / Git / PHP / WSL
"""
from __future__ import annotations

import json
import os
import re
import shutil
import subprocess
import urllib.request
import zipfile

from shared.paths import is_windows, log

# 内置兜底版本（动态解析失败时使用；2026-08 确认存在）
FALLBACK = {
    "node":   ("https://nodejs.org/dist/v24.19.0/node-v24.19.0-x64.msi", "msi"),
    "python": ("https://www.python.org/ftp/python/3.12.10/python-3.12.10-amd64.exe", "python"),
    "git":    ("https://github.com/git-for-windows/git/releases/download/v2.47.1.windows.1/Git-2.47.1-64-bit.exe", "git"),
    "php":    ("https://windows.php.net/downloads/releases/php-8.3.14-nts-Win32-vs16-x64.zip", "phpzip"),
}

# 检测命令（PATH 查找）
DETECT = {
    "node":   ["node", "node.exe"],
    "python": ["python", "python.exe"],
    "git":    ["git", "git.exe"],
    "php":    ["php", "php.exe"],
    "wsl":    ["wsl", "wsl.exe"],
}

COMPONENTS = [
    {"key": "node", "name": "Node.js（长期支持版）",
     "desc": "JavaScript 运行环境。安装器已内置一份，这里可再装系统版供其他程序使用。",
     "official": "https://nodejs.org/"},
    {"key": "python", "name": "Python 3",
     "desc": "通用编程语言，很多工具脚本需要它。静默安装到当前用户，自动加入 PATH。",
     "official": "https://www.python.org/downloads/"},
    {"key": "git", "name": "Git",
     "desc": "代码版本管理工具。开发必备，静默安装，不重启。",
     "official": "https://git-scm.com/download/win"},
    {"key": "php", "name": "PHP",
     "desc": "网站后端语言。自动下载绿色版并解压到用户目录，加入 PATH。",
     "official": "https://windows.php.net/download"},
    {"key": "wsl", "name": "WSL（Windows 的 Linux 子系统）",
     "desc": "可以在 Windows 里运行 Linux。安装需管理员权限，可能要求重启。",
     "official": "https://learn.microsoft.com/zh-cn/windows/wsl/install"},
]

DL_DIR = os.path.join(os.environ.get("TEMP", "."), "dsh-opt")


def _http_json(url: str, timeout: int = 20):
    req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req, timeout=timeout) as r:
        return json.loads(r.read().decode("utf-8", errors="replace"))


def _http_text(url: str, timeout: int = 20) -> str:
    req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req, timeout=timeout) as r:
        return r.read().decode("utf-8", errors="replace")


def _resolve_url(key: str) -> tuple[str, str] | None:
    """动态解析官方最新版下载地址，失败返回内置兜底；再失败返回 None"""
    try:
        if key == "node":
            d = _http_json("https://nodejs.org/dist/index.json")
            for item in d:
                if item.get("lts"):
                    v = item["version"]  # v24.19.0
                    return (f"https://nodejs.org/dist/{v}/node-{v}-x64.msi", "msi")
        elif key == "python":
            html = _http_text("https://www.python.org/ftp/python/")
            vers = re.findall(r'href="(\d+\.\d+\.\d+)/"', html)
            if vers:
                v = sorted(vers, key=lambda s: tuple(map(int, s.split("."))))[-1]
                return (f"https://www.python.org/ftp/python/{v}/python-{v}-amd64.exe", "python")
        elif key == "git":
            d = _http_json("https://api.github.com/repos/git-for-windows/git/releases/latest")
            tag = d.get("tag_name", "")          # v2.47.1.windows.1
            ver = tag.replace("v", "").split(".windows.")[0]  # 2.47.1
            url = (f"https://github.com/git-for-windows/git/releases/download/{tag}/"
                   f"Git-{ver}-64-bit.exe")
            return (url, "git")
        elif key == "php":
            html = _http_text("https://windows.php.net/downloads/releases/")
            vers = re.findall(r'href="(php-(8\.\d+\.\d+)-nts-Win32-vs16-x64\.zip)"', html)
            if vers:
                best = max(vers, key=lambda m: tuple(int(x) for x in m[1].split(".")))
                return ("https://windows.php.net/downloads/releases/" + best[0], "phpzip")
    except Exception as e:
        log(f"resolve {key} latest failed: {e}", "opt")
    return FALLBACK.get(key)


def is_installed(key: str) -> bool:
    if not is_windows() and key == "wsl":
        return shutil.which("wsl") is not None
    for name in DETECT.get(key, []):
        if shutil.which(name):
            return True
    # WSL 特殊：检测 wsl.exe 系统路径
    if key == "wsl":
        sysroot = os.environ.get("SystemRoot", r"C:\Windows")
        return os.path.isfile(os.path.join(sysroot, "System32", "wsl.exe"))
    return False


def _download(url: str, dest: str, progress_cb) -> bool:
    os.makedirs(DL_DIR, exist_ok=True)
    try:
        req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
        with urllib.request.urlopen(req, timeout=60) as r:
            total = int(r.headers.get("Content-Length") or 0)
            done = 0
            with open(dest, "wb") as f:
                while True:
                    chunk = r.read(1024 * 256)
                    if not chunk:
                        break
                    f.write(chunk)
                    done += len(chunk)
                    if total:
                        progress_cb(min(100, int(done * 100 / total)))
        return os.path.getsize(dest) > 0
    except Exception as e:
        log(f"download {url} failed: {e}", "opt")
        try:
            if os.path.exists(dest):
                os.remove(dest)
        except Exception:
            pass
        return False


def _run_silent(args: list[str], timeout: int = 900) -> bool:
    try:
        flags = getattr(subprocess, "CREATE_NO_WINDOW", 0)
        proc = subprocess.Popen(args, creationflags=flags)
        proc.wait(timeout=timeout)
        return proc.returncode in (0, 3010)  # 3010=成功但需重启
    except Exception as e:
        log(f"silent run failed: {args} -> {e}", "opt")
        return False


def _add_user_path(new_dir: str) -> bool:
    """把目录追加到当前用户 PATH（注册表），并广播环境变更"""
    if not is_windows():
        return False
    try:
        import winreg
        key_path = r"Environment"
        with winreg.OpenKey(winreg.HKEY_CURRENT_USER, key_path, 0,
                            winreg.KEY_READ | winreg.KEY_SET_VALUE) as key:
            try:
                cur, _ = winreg.QueryValueEx(key, "Path")
            except FileNotFoundError:
                cur = ""
        parts = [p for p in cur.split(";") if p and p.strip()]
        if new_dir not in parts:
            parts.append(new_dir)
            new_path = ";".join(parts)
            with winreg.OpenKey(winreg.HKEY_CURRENT_USER, key_path, 0,
                                winreg.KEY_SET_VALUE) as key:
                winreg.SetValueEx(key, "Path", 0, winreg.REG_EXPAND_SZ, new_path)
        return True
    except Exception as e:
        log(f"add user path failed: {e}", "opt")
        return False


def install_component(key: str, progress_cb, status_cb) -> tuple[bool, str]:
    """下载并静默安装。status_cb(阶段文字)，progress_cb(0-100)"""
    if key == "wsl":
        status_cb("正在启用 WSL 功能…（需要管理员权限）")
        sysroot = os.environ.get("SystemRoot", r"C:\Windows")
        wsl_exe = os.path.join(sysroot, "System32", "wsl.exe")
        if not os.path.isfile(wsl_exe):
            return False, "系统没有 wsl.exe，请打开微软官网按说明启用"
        ok = _run_silent([wsl_exe, "--install", "--no-distribution"], timeout=600)
        return ok, ("WSL 功能安装成功，可能需要重启电脑生效" if ok
                    else "WSL 安装未完成（可能被拒绝或需要管理员权限）")

    resolved = _resolve_url(key)
    if not resolved:
        return False, "无法获取下载地址，请点击「打开官网」手动安装"
    url, kind = resolved
    ext = ".msi" if kind == "msi" else ".exe" if kind in ("python", "git") else ".zip"
    dest = os.path.join(DL_DIR, f"{key}-latest{ext}")

    status_cb("正在下载官方安装包…")
    if not _download(url, dest, progress_cb):
        return False, "下载失败，请检查网络或点击「打开官网」手动安装"

    status_cb("正在静默安装…（无需操作，请稍候）")
    if kind == "msi":
        ok = _run_silent(["msiexec", "/i", dest, "/qn", "/norestart"])
    elif kind == "python":
        ok = _run_silent([dest, "/quiet", "InstallAllUsers=0", "PrependPath=1",
                          "Include_test=0", "Include_launcher=0"])
    elif kind == "git":
        ok = _run_silent([dest, "/VERYSILENT", "/NORESTART", "/SP-"])
    elif kind == "phpzip":
        target = os.path.join(os.environ.get("LOCALAPPDATA", os.path.expanduser("~")), "php")
        os.makedirs(target, exist_ok=True)
        try:
            with zipfile.ZipFile(dest) as z:
                z.extractall(target)
            ok = _add_user_path(os.path.join(target, os.listdir(target)[0]))
        except Exception as e:
            log(f"php zip install failed: {e}", "opt")
            ok = False
    else:
        ok = False

    # 清理安装包
    try:
        os.remove(dest)
    except Exception:
        pass

    if ok and is_installed(key):
        return True, "安装完成，新开的终端窗口即可直接使用"
    if ok:
        return True, "安装完成（未检测到命令，可能需要重开终端或重启）"
    return False, "安装未成功，请点击「打开官网」手动安装"

File: winui/DshInstaller/Core/OptComponents.cs
using System.Diagnostics;
using System.IO;
using System.IO.Compression;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.Win32;

namespace DshInstaller.Core;

public class OptComponent
{
    public string Key = "";
    public string Name = "";
    public string Desc = "";
    public string Official = "";
}

public static class OptComponents
{
    public static readonly OptComponent[] Components =
    {
        new() { Key = "node",   Name = "Node.js（长期支持版）", Desc = "JavaScript 运行环境。安装器已内置一份，这里可再装系统版供其他程序使用。", Official = "https://nodejs.org/" },
        new() { Key = "python", Name = "Python 3",             Desc = "通用编程语言，很多工具脚本需要它。静默安装到当前用户，自动加入 PATH。", Official = "https://www.python.org/downloads/" },
        new() { Key = "git",    Name = "Git",                  Desc = "代码版本管理工具，开发必备。静默安装，不重启。", Official = "https://git-scm.com/download/win" },
        new() { Key = "php",    Name = "PHP",                  Desc = "网站后端语言。自动下载绿色版并解压到用户目录，加入 PATH。", Official = "https://windows.php.net/download" },
        new() { Key = "wsl",    Name = "WSL（Windows 的 Linux 子系统）", Desc = "可以在 Windows 里运行 Linux。安装需管理员权限，可能要求重启。", Official = "https://learn.microsoft.com/zh-cn/windows/wsl/install" },
    };

    private static readonly string Dldir = Path.Combine(Path.GetTempPath(), "dsh-opt");
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromMinutes(5) };
    private static readonly Dictionary<string, (string Url, string Kind)> Fallback = new()
    {
        ["node"] = ("https://nodejs.org/dist/v24.19.0/node-v24.19.0-x64.msi", "msi"),
        ["python"] = ("https://www.python.org/ftp/python/3.12.10/python-3.12.10-amd64.exe", "python"),
        ["git"] = ("https://github.com/git-for-windows/git/releases/download/v2.47.1.windows.1/Git-2.47.1-64-bit.exe", "git"),
        ["php"] = ("https://windows.php.net/downloads/releases/php-8.3.14-nts-Win32-vs16-x64.zip", "phpzip"),
    };

    private static string? FindOnPath(string exe)
    {
        var pathVar = Environment.GetEnvironmentVariable("PATH") ?? "";
        foreach (var dir in pathVar.Split(';', StringSplitOptions.RemoveEmptyEntries))
        {
            try
            {
                var full = Path.Combine(dir.Trim('"'), exe);
                if (File.Exists(full)) return full;
            }
            catch { }
        }
        return null;
    }

    public static bool IsInstalled(string key)
    {
        switch (key)
        {
            case "node": return FindOnPath("node.exe") is not null;
            case "python": return FindOnPath("python.exe") is not null;
            case "git": return FindOnPath("git.exe") is not null;
            case "php": return FindOnPath("php.exe") is not null;
            case "wsl":
            {
                var sysroot = Environment.GetEnvironmentVariable("SystemRoot") ?? @"C:\Windows";
                return File.Exists(Path.Combine(sysroot, "System32", "wsl.exe"));
            }
            default: return false;
        }
    }

    private static (string Url, string Kind)? Resolve(string key)
    {
        try
        {
            switch (key)
            {
                case "node":
                {
                    var doc = JsonDocument.Parse(Http.GetStringAsync("https://nodejs.org/dist/index.json").Result);
                    foreach (var el in doc.RootElement.EnumerateArray())
                    {
                        if (el.TryGetProperty("lts", out var lts) && lts.ValueKind == System.Text.Json.JsonValueKind.String)
                        {
                            var v = el.GetProperty("version").GetString()!;
                            return ($"https://nodejs.org/dist/{v}/node-{v}-x64.msi", "msi");
                        }
                    }
                    break;
                }
                case "python":
                {
                    var html = Http.GetStringAsync("https://www.python.org/ftp/python/").Result;
                    var best = System.Text.RegularExpressions.Regex.Matches(html, @"href=""(\d+\.\d+\.\d+)/""")
                        .Select(m => m.Groups[1].Value)
                        .OrderByDescending(s => System.Version.Parse(s))
                        .FirstOrDefault();
                    if (best is not null)
                        return ($"https://www.python.org/ftp/python/{best}/python-{best}-amd64.exe", "python");
                    break;
                }
                case "git":
                {
                    var doc = JsonDocument.Parse(Http.GetStringAsync("https://api.github.com/repos/git-for-windows/git/releases/latest").Result);
                    var tag = doc.RootElement.GetProperty("tag_name").GetString()!;
                    var ver = tag.Replace("v", "").Split(".windows.")[0];
                    return ($"https://github.com/git-for-windows/git/releases/download/{tag}/Git-{ver}-64-bit.exe", "git");
                }
                case "php":
                {
                    var html = Http.GetStringAsync("https://windows.php.net/downloads/releases/").Result;
                    var best = System.Text.RegularExpressions.Regex.Matches(html,
                            @"href=""(php-(8\.\d+\.\d+)-nts-Win32-vs16-x64\.zip)""")
                        .Select(m => (File: m.Groups[1].Value, Ver: System.Version.Parse(m.Groups[2].Value)))
                        .OrderByDescending(x => x.Ver)
                        .FirstOrDefault();
                    if (best.File is not null)
                        return ($"https://windows.php.net/downloads/releases/{best.File}", "phpzip");
                    break;
                }
            }
        }
        catch { }
        return Fallback.TryGetValue(key, out var fb) ? fb : null;
    }

    private static bool Download(string url, string dest, IProgress<int> progress)
    {
        try
        {
            Directory.CreateDirectory(Dldir);
            using var resp = Http.GetAsync(url, HttpCompletionOption.ResponseHeadersRead).Result;
            if (!resp.IsSuccessStatusCode) return false;
            long total = resp.Content.Headers.ContentLength ?? 0;
            using var src = resp.Content.ReadAsStream();
            using var fs = new FileStream(dest, FileMode.Create, FileAccess.Write);
            var buf = new byte[256 * 1024];
            long done = 0;
            int n;
            while ((n = src.Read(buf, 0, buf.Length)) > 0)
            {
                fs.Write(buf, 0, n);
                done += n;
                if (total > 0) progress?.Report((int)(done * 100 / total));
            }
            return new FileInfo(dest).Length > 0;
        }
        catch { return false; }
    }

    private static bool RunSilent(string file, string args, int timeoutSec = 900)
    {
        try
        {
            var psi = new ProcessStartInfo(file, args)
            {
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            using var p = Process.Start(psi);
            if (p is null) return false;
            return p.WaitForExit(timeoutSec * 1000) && p.ExitCode is 0 or 3010;
        }
        catch { return false; }
    }

    private static bool AddUserPath(string dir)
    {
        try
        {
            using var key = Registry.CurrentUser.CreateSubKey("Environment");
            if (key is null) return false;
            var cur = (key.GetValue("Path") as string) ?? "";
            var parts = cur.Split(';', StringSplitOptions.RemoveEmptyEntries).Select(s => s.Trim()).ToList();
            if (!parts.Contains(dir))
            {
                parts.Add(dir);
                key.SetValue("Path", string.Join(";", parts), RegistryValueKind.ExpandString);
            }
            return true;
        }
        catch { return false; }
    }

    public static (bool ok, string msg) Install(string key, IProgress<int>? progress = null, Action<string>? status = null)
    {
        if (key == "wsl")
        {
            status?.Invoke("正在启用 WSL 功能…（需要管理员权限）");
            var sysroot = Environment.GetEnvironmentVariable("SystemRoot") ?? @"C:\Windows";
            var wslExe = Path.Combine(sysroot, "System32", "wsl.exe");
            if (!File.Exists(wslExe)) return (false, "系统没有 wsl.exe，请打开微软官网按说明启用");
            var ok = RunSilent(wslExe, "--install --no-distribution", 600);
            return ok
                ? (true, "WSL 功能安装成功，可能需要重启电脑生效")
                : (false, "WSL 安装未完成（可能被拒绝或需要管理员权限）");
        }

        var resolved = Resolve(key);
        if (resolved is null) return (false, "无法获取下载地址，请点击「打开官网」手动安装");
        var (url, kind) = resolved.Value;
        var ext = kind switch { "msi" => ".msi", "python" or "git" => ".exe", _ => ".zip" };
        var dest = Path.Combine(Dldir, $"{key}-latest{ext}");

        status?.Invoke("正在下载官方安装包…");
        if (!Download(url, dest, progress ?? new Progress<int>()))
            return (false, "下载失败，请检查网络或点击「打开官网」手动安装");

        status?.Invoke("正在静默安装…（无需操作，请稍候）");
        bool ok2 = kind switch
        {
            "msi" => RunSilent("msiexec", $"/i \"{dest}\" /qn /norestart"),
            "python" => RunSilent(dest, "/quiet InstallAllUsers=0 PrependPath=1 Include_test=0 Include_launcher=0"),
            "git" => RunSilent(dest, "/VERYSILENT /NORESTART /SP-"),
            "phpzip" => InstallPhpZip(dest),
            _ => false,
        };

        try { File.Delete(dest); } catch { }

        if (ok2 && IsInstalled(key)) return (true, "安装完成，新开的终端窗口即可直接使用");
        if (ok2) return (true, "安装完成（未检测到命令，可能需要重开终端或重启）");
        return (false, "安装未成功，请点击「打开官网」手动安装");
    }

    private static bool InstallPhpZip(string zipPath)
    {
        try
        {
            var target = Path.Combine(Environment.GetEnvironmentVariable("LOCALAPPDATA") ?? Path.GetTempPath(), "php");
            Directory.CreateDirectory(target);
            ZipFile.ExtractToDirectory(zipPath, target, overwriteFiles: true);
            var phpDir = Directory.GetDirectories(target).FirstOrDefault() ?? target;
            return AddUserPath(phpDir);
        }
        catch { return false; }
    }
}
